docs/content/develop/permadiff.md

```diff
@@ -0,0 +1,261 @@
+    return v
+}
+```
+{{< /tab >}}
+{{< /tabs >}}
+
+## API normalizes a value
+
+In cases where the API normalizes and returns a value in a simple, predictable way (such as capitalizing the value) add a diff suppress function for the field to suppress the diff.
+
+The `tpgresource` package in each provider supplies diff suppress functions for the following common cases:
+
+- `tpgresource.CaseDiffSuppress`: Suppress diffs from capitalization differences between the user's configuration and the API.
+- `tpgresource.DurationDiffSuppress`: Suppress diffs from duration format differences such as "60.0s" vs "60s". This is necessary for [`Duration`](https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#duration) API fields.
+- `tpgresource.ProjectNumberDiffSuppress`: Suppress diffs caused by the provider sending a project ID and the API returning a project number.
+
+
+{{< tabs "diff_suppress_simple" >}}
+
+{{< tab "MMv1" >}}
+```yaml
+# Use a built-in function
+diff_suppress_func: tpgresource.CaseDiffSuppress
+
+# Reference a resource-specific function
+diff_suppress_func: resourceNameFieldNameDiffSuppress
+```
+
+Define resource-specific functions in a [`custom_code.constants`](https://googlecloudplatform.github.io/magic-modules/develop/custom-code/#add-reusable-variables-and-functions) file.
+
+```go
+func resourceNameFieldNameDiffSuppress(_, old, new string, _ *schema.ResourceData) bool {
+    // Separate function for easier unit testing
+    return resourceNameFieldNameDiffSuppressLogic(old, new)
+}
+
+func resourceNameFieldNameDiffSuppressLogic(old, new) bool {
+    // Diff suppression logic. Returns true if the diff should be suppressed - that is, if the
+    // old and new values should be considered "the same".
+}
+```
+
+See [SDKv2 Schema Behaviors - DiffSuppressFunc ↗](https://developer.hashicorp.com/terraform/plugin/sdkv2/schemas/schema-behaviors#diffsuppressfunc) for more information.
+{{< /tab >}}
+{{< tab "Handwritten" >}}
+Define resource-specific functions in your service package, for example at the top of the related resource file.
+
+```go
+func resourceNameFieldNameDiffSuppress(_, old, new string, _ *schema.ResourceData) bool {
+    // Separate function for easier unit testing
+    return resourceNameFieldNameDiffSuppressLogic(old, new)
+}
+
+func resourceNameFieldNameDiffSuppressLogic(old, new) bool {
+    // Diff suppression logic. Returns true if the diff should be suppressed - that is, if the
+    // old and new values should be considered "the same".
+}
+```
+
+Reference diff suppress functions from the field definition.
+
+```go
+"field": {
+    // ...
+    DiffSuppressFunc: resourceNameFieldNameDiffSuppress,
+}
+```
+
+See [SDKv2 Schema Behaviors - DiffSuppressFunc ↗](https://developer.hashicorp.com/terraform/plugin/sdkv2/schemas/schema-behaviors#diffsuppressfunc) for more information.
+{{< /tab >}}
+{{< /tabs >}}
+
+## API does not return a field that was sent
```

Comment:
Maybe specify regardless of value- otherwise it's a default_if_empty. ignore_read means the field always causes a diff post-import so we want to minimise usage.